auto_round_extension/cuda/qlinear_tritonv2.py

```diff
@@ -155,24 +219,72 @@ def pack(self, linear, scales, zeros, g_idx=None):
-
-        out = quant_linear_fn.apply(
-            x.reshape(-1, x.shape[-1]),
-            self.qweight,
-            self.scales,
-            self.qzeros,
-            self.g_idx,
-            self.bits,
-            self.maxq,
-        )
-        out = out.half().reshape(out_shape)
-        out = out + self.bias if self.bias is not None else out
-        return out.to(x.dtype)
+        x = x.reshape(-1, x.shape[-1])
+        x_dtype = x.dtype
+        if not hasattr(self, "g_idx"):
+                self.g_idx = torch.tensor(
+                    [i // self.group_size for i in range(self.infeatures)], dtype=torch.int32
+                ).to(self.qweight.device)
+        if self.bits in [2, 4, 8]:
+            quant_linear_fn = QuantLinearFunction
+            out = quant_linear_fn.apply(
+                x,
+                self.qweight,
+                self.scales,
+                self.qzeros,
+                self.g_idx,
+                self.bits,
+                self.maxq,
+            )
+            out = out.half().reshape(out_shape)
+            out = out + self.bias if self.bias is not None else out
+            return out.to(x.dtype)
+        elif self.bits == 3:
+            if self.wf.device != self.qzeros.device:
+                self.wf = self.wf.to(self.qzeros.device)
+            zeros = self.qzeros.reshape(self.qzeros.shape[0], self.qzeros.shape[1] // 3, 3, 1).expand(
+                -1, -1, -1, 12
+            )
+            zeros = zeros >> self.wf.unsqueeze(0)
+            zeros[:, :, 0, 10] = (zeros[:, :, 0, 10] & 0x3) | ((zeros[:, :, 1, 0] << 2) & 0x4)
+            zeros[:, :, 1, 11] = (zeros[:, :, 1, 11] & 0x1) | ((zeros[:, :, 2, 0] << 1) & 0x6)
+            zeros = zeros & 0x7
+            zeros = torch.cat(
+                [zeros[:, :, 0, :11], zeros[:, :, 1, 1:12], zeros[:, :, 2, 1:11]],
+                dim=2,
+            )
+
+            zeros = zeros.reshape(self.scales.shape)
+
+            weight = self.qweight.reshape(self.qweight.shape[0] // 3, 3, 1, self.qweight.shape[1]).expand(
+                -1, -1, 12, -1
+            )
+            weight = (weight >> self.wf.unsqueeze(-1)) & 0x7
+            weight[:, 0, 10] = (weight[:, 0, 10] & 0x3) | ((weight[:, 1, 0] << 2) & 0x4)
+            weight[:, 1, 11] = (weight[:, 1, 11] & 0x1) | ((weight[:, 2, 0] << 1) & 0x6)
+            weight = weight & 0x7
+            weight = torch.cat([weight[:, 0, :11], weight[:, 1, 1:12], weight[:, 2, 1:11]], dim=1)
+
+            weight = weight.reshape(weight.shape[0] * weight.shape[1], weight.shape[2])
+            num_itr = self.g_idx.shape[0] // x.shape[-1]
+            if num_itr == 1:
+                weights = self.scales[self.g_idx.long()] * (weight - zeros[self.g_idx.long()])
+            else:
+                num_dim = self.g_idx.shape[0] // num_itr
+                weights = []
+                for i in range(num_itr):
+                    scale_i = self.scales[:, i * num_dim : (i + 1) * num_dim]
```

Comment:
Please prioritize optimizing this section, specifically, replace the for loop with tensor operations, as we're using a dummy g_idx. The current implementation is significantly slowing down inference.